packages/blob/src/put.ts

```diff
@@ -42,23 +42,25 @@ export function createPutMethod<
   getToken?: (pathname: string, options: T) => Promise<string>;
   extraChecks?: (options: T) => void;
 }) {
-  return async function put(
-    pathname: string,
-    body:
-      | string
-      | Readable
-      | Blob
-      | ArrayBuffer
-      | FormData
-      | ReadableStream
-      | File,
+  return async function put<TPath extends string>(
+    pathname: TPath,
+    body: TPath extends `${infer _rest}/`
```

Comment:
Is this compatible with TS 4 too? Just wondering I never saw it yet. Can you show a screenshot/video of the DX in VSCode?

Reply:
Video:

https://github.com/vercel/storage/assets/41523937/0062f813-6110-4154-b67d-185f29b3cb63

Reply:
i think the `infer` keyword was already part of TS 4 so it should be compatible

Reply:
also i think this can be done without infer... stay tuned 😅

Reply:
ok, here is my second shot at this. It's less verbose for the user since no explicit `undefined` is needed but it might be too much magic. 
Also the error is not really helpful when you switch from file to folder and ts thinks the options object is the body

https://github.com/vercel/storage/assets/41523937/22a35098-5c8f-45bc-8134-86a33be80546

Reply:
Pretty cool, thank you!